realm-jni/src/io_realm_internal_table.cpp

```diff
@@ -1500,3 +1505,33 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_Table_nativeSetPrimaryKey(
     } CATCH_STD()
     return 0;
 }
+
+// Fixes interop issue with Cocoa Realm where the Primary Key table had different types.
+// This affects:
+// - All Realms created by Cocoa and used by Realm-android up to 0.80.1
+// - All Realms created by Realm-Android 0.80.1 and below
+// See https://github.com/realm/realm-java/issues/1059
+JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeMigratePrimaryKeyTableIfNeeded
+    (JNIEnv* env, jobject, jlong groupNativePtr, jlong privateKeyTableNativePtr)
+{
+    Group* group = G(groupNativePtr);
+    Table* pk_table = TBL(privateKeyTableNativePtr);
+    if (pk_table->get_column_type(io_realm_internal_Table_PRIMARY_KEY_FIELD_COLUMN_INDEX) == type_Int) {
+        StringData tmp_col_name = StringData("tmp_field_name");
+        size_t tmp_col_ndx = pk_table->add_column(DataType(type_String), tmp_col_name);
+
+        // Create tmp string column with field name instead of column index
+        size_t number_of_rows = pk_table->size();
+        for (size_t row_ndx = 0; row_ndx < number_of_rows; row_ndx++) {
+            StringData table_name = pk_table->get_string(io_realm_internal_Table_PRIMARY_KEY_CLASS_COLUMN_INDEX, row_ndx);
+            size_t col_ndx = static_cast<size_t>(pk_table->get_int(io_realm_internal_Table_PRIMARY_KEY_FIELD_COLUMN_INDEX, row_ndx));
+            StringData col_name = group->get_table(table_name)->get_column_name(col_ndx);
+            pk_table->set_string(tmp_col_ndx, row_ndx, col_name);
+        }
+
+        // Delete old int column, and rename tmp column to same name
+        // The column index for the renamed column will then be the same as the deleted old column
```

Comment:
This is not generally true but you know the exact layout of the table (two columns: string, integer). It might be worth describing the old layout and the new layout (string, string) in a comment at the top of the function.